### PR TITLE
Intgrate spec for views

### PR DESCRIPTION
**Pull request by souad988**
solved  N+1 problem when fetching all posts and their comments for a user.
**after**
Capture d’écran du 2023-10-02 14-35-12
**before**
Capture d’écran du 2023-10-02 14-35-01

Use Capybara to write integration tests for each view in the project:

- User index page
- user show page
- user post index
- user post show

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -2,8 +2,10 @@ class PostsController < ApplicationController
   def index
     if params[:user_id].present?
       # List posts of a specific user
+
       @user = User.find(params[:user_id])
-      @posts = @user.posts
+      # @posts = @user.posts
```
Comment (mikethreels) [inline app/controllers/posts_controller.rb]: - [ ] Please remove any commented-out code it makes your codebase look unorganised and harder to read.

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -2,8 +2,10 @@ class PostsController < ApplicationController
...
+      @posts = Post.includes(:author).where(author: params[:user_id])
```
Comment (mikethreels) [inline app/controllers/posts_controller.rb]: - [ ] To solve the N+1 problem you're supposed to get the comments when fetching the posts, not the author.
you can archive this by writing this line like: `@posts = @user.posts.includes(:comments)`

File: spec/views/user_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    expect(page).to have_content('User 1')
+    expect(page).to have_content('User 2')
+
+    # Verify that the profile picture for each user is displayed
+    expect(page).to have_css("img[src*='user1.jpg']")
+    expect(page).to have_css("img[src*='user2.jpg']")
+
+    # Verify that the number of posts each user has written is displayed
+    # expect(page).to have_content('User 1 has written 0 posts.')
+    # expect(page).to have_content('User 2 has written 0 posts.')
```
Comment (mikethreels) [inline spec/views/user_index_spec.rb]: - [ ] Please remove any commented-out code it makes your codebase look unorganised and harder to read.

File: spec/views/user_posts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    expect(page).to have_content('Post 1')
+    expect(page).to have_content('This is post 1')
+    expect(page).to have_content('Comment 1')
+    expect(page).to have_content('comments: 5')
+    expect(page).to have_content('likes: 10')
+  end
+  scenario 'redirects to the post\'s show page when clicked' do
+    click_link 'Post 1'
+    expect(page).to have_current_path(user_post_path(user, user.posts.first))
+  end
```
Comment (mikethreels) [inline spec/views/user_posts_spec.rb]: - [ ] in the user_post_index tests there should be a Test that asserts you can see a section for pagination if there are more posts than fit on the view.

File: spec/views/user_posts_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    expect(page).to have_content('Test User')
+  end
+  scenario 'displays the body content ,first comments on a post how many comments and how many likes' do
+    expect(page).to have_content('posts count: 10')
+    expect(page).to have_content('Post 1')
+    expect(page).to have_content('This is post 1')
+    expect(page).to have_content('Comment 1')
+    expect(page).to have_content('comments: 5')
+    expect(page).to have_content('likes: 10')
+  end
```
Comment (mikethreels) [inline spec/views/user_posts_spec.rb]: - [ ] [OPTIONAL] In the user index and user show test you divided each test with a description which is best practice when writing tests, I highly recommend you do this for all the tests.

File: spec/views/user_post_show_spec.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    expect(page).to have_content('Likes:10')
+  end
+
+  scenario 'displays the post body' do
+    expect(page).to have_content('This is a test post')
+  end
+
+  scenario 'displays the username of each commentor' do
+    expect(page).to have_content('Test User:', count: 5)
+  end
```
Comment (mikethreels) [inline spec/views/user_post_show_spec.rb]: - [ ] in the user_post_show spec there should be a Test that asserts you can see the comment each commentor left.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- good job writing the tests.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (souad988) [reply app/controllers/posts_controller.rb]: thanks @mikethreels for the help

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (Lucas-Masaba) [reply app/controllers/posts_controller.rb]: - [ ] Please remove this commented code as requested by the previous reviewer.

File: spec/views/user_index_spec.rb (hunk 3 above)
Comment (Lucas-Masaba) [reply spec/views/user_index_spec.rb]: - [ ] Please remove this commented code as requested by the previous reviewer.

**Review by Lucas-Masaba — CHANGES_REQUESTED** [review]
# Hello @migo1 && @souad988  👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
almost-there

 ## Highlights 🍾 
- Linter checks passing :heavy_check_mark: 
- Added tests and they are all passing :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏 💻 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Lucas-Masaba in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mahbubul14 — APPROVED** [review]
# Project Approved :trophy: :tada: 

Hi @souad988 , 

Your project is complete!  You have done it exceptionally well. it's time to merge it :100: 
Congratulations! 🎉 

### To Highlight :+1: 

:heavy_check_mark: No linter error.
:heavy_check_mark: Descriptive readme file.
:heavy_check_mark: Well-implemented suggested changes.

### Optional suggestions :arrow_down_small: 

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better.

* N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification. You can also connect with me on [slack](https://microverse-students.slack.com/team/U01J39GL614)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Pull request merged**